packages/datadog-plugin-undici/src/index.js

```diff
@@ -0,0 +1,341 @@
+'use strict'
+
+const url = require('url')
+const log = require('../../dd-trace/src/log')
+const tags = require('../../../ext/tags')
+const kinds = require('../../../ext/kinds')
+const formats = require('../../../ext/formats')
+const urlFilter = require('../../dd-trace/src/plugins/util/urlfilter')
+const analyticsSampler = require('../../dd-trace/src/analytics_sampler')
+const { AsyncResource, AsyncLocalStorage } = require('async_hooks')
+
+const HTTP_HEADERS = formats.HTTP_HEADERS
+const HTTP_STATUS_CODE = tags.HTTP_STATUS_CODE
+const HTTP_REQUEST_HEADERS = tags.HTTP_REQUEST_HEADERS
+const HTTP_RESPONSE_HEADERS = tags.HTTP_RESPONSE_HEADERS
+const SPAN_KIND = tags.SPAN_KIND
+const CLIENT = kinds.CLIENT
+
+const asyncLocalStorage = new AsyncLocalStorage()
+
+function addError (span, error) {
+  span.addTags({
+    'error.type': error.name,
+    'error.msg': error.message,
+    'error.stack': error.stack
+  })
+  return error
+}
+
+function parseHeaders (headers) {
+  const pairs = headers
+    .split('\r\n')
+    .map((r) => r.split(':').map((str) => str.trim()))
+
+  const object = {}
+  pairs.forEach(([key, value]) => {
+    key = key.toLowerCase()
+    if (!value) {
+      return
+    }
+    if (object[key]) {
+      if (!Array.isArray(object[key])) {
+        object[key] = [object[key], value]
+      } else {
+        object[key].push(value)
+      }
+    } else {
+      object[key] = value
+    }
+  })
+  return object
+}
+const channels = {
+  requestChannel: undefined,
+  headersChannel: undefined,
+  errorChannel: undefined
+}
+
+function diagnostics (tracer, config) {
+  let diagnosticsChannel
+  try {
+    diagnosticsChannel = require('diagnostics_channel')
+  } catch (e) {
+    log.error(
+      "Unable to configure undici, cannot require 'diagnostics_channel'"
+    )
+    return () => {}
+  }
+  config = normalizeConfig(tracer, config)
+
+  channels.requestChannel = diagnosticsChannel.channel('undici:request:create')
+  channels.headersChannel = diagnosticsChannel.channel(
+    'undici:request:headers'
+  )
+  channels.errorChannel = diagnosticsChannel.channel('undici:request:error')
+
+  channels.requestChannel.subscribe(handleRequestCreate)
+  channels.errorChannel.subscribe(handleRequestError)
+  channels.headersChannel.subscribe(handleRequestHeaders)
+
+  const requestSpansMap = new WeakMap()
+
+  function handleRequestCreate ({ request }) {
+    const method = (request.method || 'GET').toUpperCase()
+
+    const path = request.path ? request.path.split(/[?#]/)[0] : '/'
+    const uri = `${request.origin}${path}`
+
+    const span = asyncLocalStorage.getStore()
+    if (span) {
+      span.addTags({
+        'resource.name': method,
+        'span.type': 'http',
+        'http.method': method,
+        'http.url': uri,
+        'service.name': getServiceName(tracer, config, request)
+      })
+      requestSpansMap.set(request, span)
```

Comment:
we keep the request in the WeakMap as for some reason, they're not available in the subsequent diagnostics channels events. Which makes me wonder why? and also if we can rely on this span to exist in the `undici:request:create` event

Reply:
Can you open an issue about this? Or a PR to add them? saving the WeakMap will greatly reduce overhead.

Reply:
Not sure what you mean. Is that more efficient to store through the weak map than the async context? 
or would you want something on the request object to add external context in the form of a weak map?